hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/datanode/DataXceiver.java

```diff
@@ -1122,6 +1124,7 @@ public void copyBlock(final ExtendedBlock block,
       datanode.metrics.incrBytesRead((int) read);
       datanode.metrics.incrBlocksRead();
       datanode.metrics.incrTotalReadTime(duration);
+      datanode.metrics.addReadTransferRateMBs(DFSUtil.transferRateMBs(read, duration));
```

Comment:
You will be publishing the metrics, including the -1 here. Assuming duration is 0, I would expect bytes read to be 0 as well (only instance with sub millisecond transfers should be when there's an error prior to the transfer). 

Bytes Read 0, Duration 0: Return 0.
Bytes Read N, duration 0: Exception which we log.
Bytes Read N, duration K: Normal

Reply:
I was thinking we can publish -1 for the error scenarios which can be visualized from the graphs as well.

Reply:
updated to log for cases when we get unexpected data.
1) Bytes read is negative
2) Duration is negative or 0.
Since we are logging these cases not emitting the data to metrics.